contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,952 @@
+                .getAmountsForLiquidity(
+                    sqrtRatioX96TickClosestToParity, // sqrtRatioX96
+                    sqrtRatioX96Tick0, // sqrtRatioAX96
+                    sqrtRatioX96Tick1, // sqrtRatioBX96
+                    liquidity
+                );
+
+            require(wethAmount == 0, "Non zero wethAmount");
+            underlyingAssets = oethbAmount;
+        }
+    }
+
+    /**
+     * @notice Deposit all supported assets in this strategy contract to the platform
+     */
+    function depositAll() external override onlyVault nonReentrant {
+        _deposit(WETH, IERC20(WETH).balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Withdraw an `amount` of assets from the platform and
+     * send to the `_recipient`.
+     * @param _recipient         Address to which the asset should be sent
+     * @param _asset             Address of the asset
+     * @param _amount            Units of asset to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        require(_asset == WETH, "Unsupported asset");
+        require(_recipient == vaultAddress, "Only withdraw to vault allowed");
+
+        uint256 wethBalance = IERC20(WETH).balanceOf(address(this));
+        if (wethBalance < _amount) {
+            uint256 additionalWethRequired = _amount - wethBalance;
+            (uint256 wethInThePool, ) = getPositionPrincipal();
+
+            if (wethInThePool < additionalWethRequired) {
+                revert NotEnoughWethLiquidity(
+                    wethInThePool,
+                    additionalWethRequired
+                );
+            }
+
+            uint256 shareOfWethToRemove = additionalWethRequired.divPrecisely(
+                wethInThePool
+            ) + 1;
+            _removePartialLiquidity(shareOfWethToRemove);
+        }
+
+        // burn remaining OETHb
+        _burnOethbOnTheContract();
+        _withdraw(_recipient, _amount);
+    }
+
+    /**
+     * @notice Withdraw all supported assets from platform and
+     * sends to the OToken's Vault.
+     */
+    function withdrawAll() external override onlyVault nonReentrant {
+        _removeAllLiquidity();
+
+        uint256 balance = IERC20(WETH).balanceOf(address(this));
+        if (balance > 0) {
+            _withdraw(vaultAddress, balance);
+        }
+        // burn remaining OETHb
+        _burnOethbOnTheContract();
+    }
+
+    function _withdraw(address _recipient, uint256 _amount) internal {
+        require(_amount > 0, "Must withdraw something");
+        require(_recipient != address(0), "Must specify recipient");
+
+        IERC20(WETH).safeTransfer(_recipient, _amount);
+        emit Withdrawal(WETH, address(0), _amount);
+    }
+
+    /**
+     * @dev Collect the AERO token from the gauge
+     */
+    function _collectRewardTokens() internal override {
+        clGauge.getReward(tokenId);
```

Comment:
🟠 This assumes that there's always liquidity on the pool. Might not always be the case. It works for now, since the strategist might know the liquidity position. However, when do a full-fledged Harvester contract, it might be an annoyance for the farmers if this call fails randomly on them